From 5335c7e1d0710d2e2dd15203ac3901815df53509 Mon Sep 17 00:00:00 2001
From: Liu Ying <Ying.liu@freescale.com>
Date: Tue, 19 Feb 2013 15:07:43 +0800
Subject: [PATCH 106/231] ENGR00243314 MX6 Sabreauto:Setup ipu di parent clock
 for HDMI case

We support smooth UI transition from Uboot splashscreen to Android
kernel for IPU1 DI1, so we don't set ipu1 di1 clock's parent clock
to be pll5 clock when kernel initializes clock tree. However,
sometimes, by using the same uboot and kernel, some platforms may
use ipu1 di1 to drive HDMI(e.g., MX6DL Sabreauto), which makes HDMI
pixel clock tree wrong. This patch sets the clock tree explicitly
in HDMI initializing hook function to address the issue. This patch
is for Android kernel only.

Signed-off-by: Liu Ying <Ying.Liu@freescale.com>
---
 arch/arm/mach-mx6/board-mx6q_sabreauto.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/arch/arm/mach-mx6/board-mx6q_sabreauto.c b/arch/arm/mach-mx6/board-mx6q_sabreauto.c
index 0ed7f15..ee326dd 100644
--- a/arch/arm/mach-mx6/board-mx6q_sabreauto.c
+++ b/arch/arm/mach-mx6/board-mx6q_sabreauto.c
@@ -929,6 +929,8 @@ static struct ipuv3_fb_platform_data sabr_fb_data[] = {
 static void hdmi_init(int ipu_id, int disp_id)
 {
 	int hdmi_mux_setting;
+	char ipu_di_clk[] = "ipu1_di0_clk";
+	struct clk *di_clk, *pll5_clk;
 
 	if ((ipu_id > 1) || (ipu_id < 0)) {
 		printk(KERN_ERR"Invalid IPU select for HDMI: %d. Set to 0\n",
@@ -951,6 +953,16 @@ static void hdmi_init(int ipu_id, int disp_id)
 	/* Set HDMI event as SDMA event2 while Chip version later than TO1.2 */
 	if (hdmi_SDMA_check())
 		mxc_iomux_set_gpr_register(0, 0, 1, 1);
+
+	ipu_di_clk[3] += ipu_id;
+	ipu_di_clk[7] += disp_id;
+	di_clk = clk_get(NULL, ipu_di_clk);
+	if (IS_ERR(di_clk))
+		printk(KERN_ERR "Cannot get %s clock\n", ipu_di_clk);
+	pll5_clk = clk_get(NULL, "pll5");
+	if (IS_ERR(pll5_clk))
+		printk(KERN_ERR "Cannot get pll5 clock\n");
+	clk_set_parent(di_clk, pll5_clk);
 }
 
 /* On mx6x sabreauto board i2c2 iomux with hdmi ddc,
-- 
1.7.10.4

